src/pybind/mgr/volumes/fs/stats_util.py

```diff
@@ -57,3 +65,114 @@ def get_stats(src_path, dst_path, fsh, human=True):
+        ce_path = os_path_join(clone_index_path, clone_entry)
+        # XXX: This may raise ObjectNotFound exception. As soon as cloning is
+        # finished, clone entry is deleted by cloner thread. This exception is
+        # handled in _get_info_for_all_clones().
+        ci.dst_path = fsh.readlink(ce_path, 4096).decode('utf-8')
+
+        ci.dst_grp_name, ci.dst_sv_name = \
+            resolve_grp_and_sv_names(self.vol_spec, ci.dst_path)
+
+    def _get_clone_src_info(self, fsh, ci):
+        with open_clone_sv_pair_in_vol(
+                self.volclient, self.vol_spec, ci.volname, ci.dst_grp_name,
+                ci.dst_sv_name) as (dst_sv, src_sv, snap_name):
+            ci.src_grp_name = src_sv.group_name
+            ci.src_sv_name = src_sv.subvolname
+            ci.src_path = src_sv.snapshot_data_path(snap_name)
+
+    def _get_info_for_all_clones(self):
+        clones = []
+
+        volnames = list_volumes(self.volclient.mgr)
+        for volname in volnames:
+            with open_volume_lockless(self.volclient, volname) as fsh:
+                with open_clone_index(fsh, self.vol_spec) as clone_index:
+                    clone_index_path = clone_index.path
+                    clone_entries = clone_index.list_entries()
+
+            for ce in clone_entries:
+                ci = CloneInfo()
+                ci.volname = volname
+
+                try:
+                    self._get_clone_dst_info(fsh, ci, ce, clone_index_path)
+                    self._get_clone_src_info(fsh, ci)
+                    if ci.src_path is None or ci.dst_path is None:
+                        continue
+                # clone entry went missing, it was removed because cloning has
+                # finished.
+                except ObjectNotFound:
+                    continue
+
+                clones.append(ci)
+
+        return clones if clones else []
+
+    def _update_progress_bar(self):
+        assert self.pev_id is not None
+        clone = self._get_info_for_all_clones()
+        if not clone:
+            self._finish()
+            return
+
+        with open_volume_lockless(self.volclient, clone.volname) as fsh:
+            _, _, percent = get_amount_copied(clone.src_path, clone.dst_path,
+                                          fsh)
+
+        # progress module takes progress as a fraction between 0.0 to 1.0.
+        progress_fraction = percent / 100
+        msg = f'Subvolume "{self.dst_svname}" has been {percent}% cloned'
+
+        self.volclient.mgr.remote('progress', 'update', ev_id=self.pev_id,
+                           ev_msg=msg, ev_progress=progress_fraction,
+                           refs=['mds', 'clone'], add_to_ceph_s=True)
+
+        if progress_fraction == 1.0:
+            self._finish()
+
+    def _finish(self):
+        assert self.pev_id is not None
+
+        self.volclient.mgr.remote('progress', 'complete', self.pev_id)
+        self.pev_id = None
+
+        self.update_task.finished.set()
```

Comment:
>I suggested keeping the timer task running and only terminate after a certain interval of inactivity. Have you thought about that?

That is the case. See `_wait_and_finish()`.

Reply:
I see that in the later commit but this comment was in one of the earlier commits. Better to squash IMO.

Reply:
Previous commit is titled `show progress bar for ongoing+pending clones`, IMO it is not suited to squash into a commit that is logical different and incoherent with this change,

Reply:
Then you'd need to fold the relevant hunk into the previous commit that introduces the timer task. You are mixing things up @rishabh-d-dave

Reply:
The snippet above doesn't wait and finish, it simply finishes without waiting. The commit that adds the feature to wait before finishing doesn't contain any other patches. See -

```
diff --git a/src/pybind/mgr/volumes/fs/stats_util.py b/src/pybind/mgr/volumes/fs/stats_util.py
index 232a48a0d29..a74585fe759 100644
--- a/src/pybind/mgr/volumes/fs/stats_util.py
+++ b/src/pybind/mgr/volumes/fs/stats_util.py
@@ -260,16 +260,66 @@ class CloneProgressReporter:
         '''
         clones = self._get_info_for_all_clones()
         if not clones:
-            self._finish()
+            self._wait_and_finish()
             return
 
         self._update_ongoing_progress_bar(clones)
         self._update_onpen_progress_bar(clones)
 
+    def _set_waiting_period_msg(self, time_left):
+        avg_progress_fraction = 1.0
+
+        msg_on = (f'All cloning finished, waiting {time_left} seconds before '
+                'exiting')
+        self._update_progress_bar_event(
+            ev_id=self.on_pev_id, ev_msg=msg_on,
+            ev_progress_fraction=avg_progress_fraction)
+
+        if self.show_onpen_bar:
+            msg_onpen = (f'No pending clones left, waiting {time_left} '
+                          'seconds before exiting')
+            self._update_progress_bar_event(
+                ev_id=self.onpen_pev_id, ev_msg=msg_onpen,
+                ev_progress_fraction=avg_progress_fraction)
+
+    def _wait_and_finish(self):
+        '''
+        All cloning has been finished. Wait for some time before terminating
+        the/this thread updating and deleting the objects for progress bar
+        events. In case new clones are launched, all these resources can be
+        reused.
+        '''
+        log.info('waiting for new clones before terminating this thread.')
+
+        interval = 1
+        wait_limit = 30
+        start_time = time()
+        cur_time = start_time
+        time_elapsed = cur_time - start_time
+        time_left = wait_limit - time_elapsed
+
+        self._set_waiting_period_msg(time_left)
+
+        while time_elapsed < wait_limit:
+            clones = self._get_info_for_all_clones()
+            if len(clones):
+                return
+
+            sleep(interval)
+            cur_time += interval
+            time_elapsed = cur_time - start_time
+            time_left = wait_limit - time_elapsed
+
+            self._set_waiting_period_msg(time_left)
+
+        log.info('no new clones were launched during waiting period, '
+                 'terminating this thread and related resources now.')
+        self._finish()
+
     def _finish(self):
         '''
-        All cloning has been finished. Remove progress bars from "ceph -s"
-        output.
+        All cloning has been finished and enough waiting has been done for new
+        clones to appear. Remove progress bars from "ceph -s" output.
         '''
         assert self.on_pev_id is not None
         assert self.onpen_pev_id is not None
```

Reply:
As per our discussion elsewhere, this feature will implemented on a separated PR, marking this thread as resolved.